### ISSUE TITLE
java.lang.OutOfMemoryError: : Java Heap Space

### ISSUE THREAD

**Issue by donnib** [rtyley/bfg-repo-cleaner#89]
mbp:GitMigration$ java -jar bfg-1.12.3.jar --delete-files "EnhancedLoggingCPB compare.xml" gitRepo.git

Using repo : /Users/mihai/GitMigration/gitRepo.git

Found 8781 objects to protect
Found 100 commit-pointing refs : HEAD, refs/heads/master, refs/remotes/origin/2250NMT, ...

Protected commits
-----------------

These are your protected commits, and so their contents will NOT be altered:

 * commit 3b234066 (protected by 'HEAD')
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOfRange(Arrays.java:3209)
	at java.lang.String.<init>(String.java:215)
	at java.lang.StringBuilder.toString(StringBuilder.java:430)
	at sun.net.www.protocol.jar.Handler.parseContextSpec(Handler.java:120)
	at sun.net.www.protocol.jar.Handler.parseURL(Handler.java:66)
	at java.net.URL.<init>(URL.java:596)
	at java.net.URL.<init>(URL.java:464)
	at sun.misc.URLClassPath$JarLoader.checkResource(URLClassPath.java:690)
	at sun.misc.URLClassPath$JarLoader.getResource(URLClassPath.java:775)
	at sun.misc.URLClassPath.getResource(URLClassPath.java:173)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:194)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:190)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:306)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:301)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:247)
	at org.eclipse.jgit.internal.storage.file.PackFile.load(PackFile.java:800)
	at org.eclipse.jgit.internal.storage.file.PackFile.get(PackFile.java:257)
	at org.eclipse.jgit.internal.storage.file.ObjectDirectory.openPackedObject(ObjectDirectory.java:411)
	at org.eclipse.jgit.internal.storage.file.ObjectDirectory.openPackedFromSelfOrAlternate(ObjectDirectory.java:380)
	at org.eclipse.jgit.internal.storage.file.ObjectDirectory.openObject(ObjectDirectory.java:372)
	at org.eclipse.jgit.internal.storage.file.WindowCursor.open(WindowCursor.java:145)
	at org.eclipse.jgit.revwalk.RevWalk.getCachedBytes(RevWalk.java:883)
	at org.eclipse.jgit.revwalk.RevCommit.parseHeaders(RevCommit.java:145)
	at org.eclipse.jgit.revwalk.PendingGenerator.next(PendingGenerator.java:148)
	at org.eclipse.jgit.revwalk.TopoSortGenerator.<init>(TopoSortGenerator.java:78)
	at org.eclipse.jgit.revwalk.StartGenerator.next(StartGenerator.java:168)
	at org.eclipse.jgit.revwalk.RevWalk.next(RevWalk.java:421)
	at org.eclipse.jgit.revwalk.RevWalk.iterator(RevWalk.java:1257)
	at scala.collection.convert.Wrappers$JIterableWrapper.iterator(Wrappers.scala:54)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)

this is a 2.8Gb repository. I am running on a Mac

**Comment by rtyley** [comment]
Can you share the repo with me?

**Comment by donnib** [comment]
It's a 2.8 gb repo with all our source code inside our company, unfortunnately i am not allowed to share it :(

**Comment by rtyley** [comment]
> unfortunnately i am not allowed to share it :(

I get that a lot! :-/ Oh well. So, although the stack trace is fairly confusing (as low-memory errors often are, because the last-straw that's reported isn't necessarily the thing that's consumed the lion's share of your memory), the important thing is the `java.lang.OutOfMemoryError: Java heap space` at the top. The process ran out of memory, and you can try allowing greater memory consumption with the `-Xmx` flag.

So you would now do:

```
$ java -Xmx2048m -jar bfg-1.12.3.jar --delete-files "EnhancedLoggingCPB compare.xml" gitRepo.git
```

..try a few different values for that `2048m` - let me know how it goes!

Also, if you could send me the output of: `git verify-pack -v .git/objects/pack/pack-*pack` - executed on the original repo, that would be interesting, to get an idea of the object sizes in your repo.

**Comment by rtyley** [comment]
Has adding `-Xmx2048m` made any difference?

**Comment by donnib** [comment]
Yes indeed i passed that problem but now another problem, it's very slow. It started really fast then went do beeing 10 revision a sec speed which seems slower than filter which it should be faster than. Any idea what causes that ? Should i open a separete case for this ?

**Comment by rtyley** [comment]
That sounds like a garbage collectino issue- try increasing the memory allocated- to 3072m for instance. You could also try ensuring that the min size of the heap is identical to the max, to avoid aggressive reclaiming of space, eg `-Xmx3072m -Xmx3072m`

**Comment by donnib** [comment]
I tried 3072 and it had no impact. 42000commits git took 20 min to delete a
file. I ran it on macbook with ssd and 8g ram.

**Comment by rtyley** [comment]
20 minutes is quite slow, but bearable. You know that you can delete [more than one file or folder](http://stackoverflow.com/a/21636596/438886) in one BFG run, right?

**Comment by donnib** [comment]
No, i did not know, i'll check that option, so you are saying that 20min is probably right ? Should i expect it to be faster on this size of repo ? As mentioned above the repo has 42000commits and when i mirror clone it, it's 1.5Gb.

**Comment by donnib** [comment]
Here is the output to 28%, it was fast to about 15-20% then it became very slow. I turned on the verbose output for Garbage Collection, maybe it helps : 

mihais-mbp:GitMigration mihai$ java -Xmx6072m -Xms6072m -verbose:gc -jar bfg-1.12.3.jar --delete-files "EnhancedLoggingCPB compare.xml" gitrepo1.git

Using repo : /Users/mihai/GitMigration/gitrepo1.git

[GC 67914K->30228K(6209216K), 0.0172871 secs]
[GC 98148K->68961K(6209216K), 0.0257528 secs]
[GC 137121K->106053K(6209216K), 0.0270529 secs]
Found 8781 objects to protect
Found 100 commit-pointing refs : HEAD, refs/heads/master, refs/remotes/origin/2250NMT, ...

Protected commits
-----------------

These are your protected commits, and so their contents will NOT be altered:

 * commit 3b234066 (protected by 'HEAD')
[GC 174213K->123101K(6209216K), 0.0279473 secs]

Cleaning
--------

Found 46121 commits
Cleaning commits:         0% (    1/46121)[GC 191261K->146868K(6209216K), 0.0341119 secs]
[GC 215028K->157750K(6209216K), 0.0159930 secs]
Cleaning commits:         1% (  462/46121)[GC 225918K->157777K(6209216K), 0.0041619 secs]
Cleaning commits:         1% (  531/46121)[GC 225937K->162403K(6209216K), 0.0092295 secs]
Cleaning commits:         2% (  923/46121)[GC 230567K->163720K(6209216K), 0.0040866 secs]
[GC 164576K->167441K(6209216K), 0.0080013 secs]
[GC 235601K->166065K(6209216K), 0.0046173 secs]
Cleaning commits:         3% ( 1384/46121)[GC 234233K->169964K(6209216K), 0.0059775 secs]
[GC 170161K->170767K(6209216K), 0.0051760 secs]
[GC 238936K->171791K(6209216K), 0.0057276 secs]
Cleaning commits:         4% ( 1845/46121)[GC 239951K->174727K(6209216K), 0.0062731 secs]
Cleaning commits:         4% ( 2097/46121)[GC 242887K->176821K(6209216K), 0.0099764 secs]
Cleaning commits:         5% ( 2307/46121)[GC 244981K->178648K(6209216K), 0.0087745 secs]
[GC 246808K->180705K(6209216K), 0.0090173 secs]
Cleaning commits:         6% ( 2768/46121)[GC 248865K->184218K(6209216K), 0.0102341 secs]
Cleaning commits:         6% ( 3004/46121)[GC 252378K->183912K(6209216K), 0.0050080 secs]
Cleaning commits:         7% ( 3229/46121)[GC 252072K->187152K(6209216K), 0.0064491 secs]
Cleaning commits:         8% ( 3690/46121)[GC 255312K->189732K(6209216K), 0.0083462 secs]
Cleaning commits:         8% ( 3884/46121)[GC 257898K->191852K(6209216K), 0.0069964 secs]
[GC 192323K->192355K(6209216K), 0.0060220 secs]
Cleaning commits:         9% ( 4151/46121)[GC 260515K->192575K(6209216K), 0.0053615 secs]
Cleaning commits:        10% ( 4613/46121)[GC 260735K->194313K(6209216K), 0.0070907 secs]
Cleaning commits:        11% ( 5074/46121)[GC 262473K->197383K(6209216K), 0.0091967 secs]
[GC 265543K->200450K(6209216K), 0.0121115 secs]
Cleaning commits:        12% ( 5535/46121)[GC 268610K->202584K(6209216K), 0.0060371 secs]
Cleaning commits:        13% ( 5996/46121)[GC 270744K->206398K(6209216K), 0.0092566 secs]
Cleaning commits:        14% ( 6457/46121)[GC 274558K->206161K(6209216K), 0.0051969 secs]
Cleaning commits:        15% ( 7040/46121)[GC 274321K->208789K(6209216K), 0.0084924 secs]
Cleaning commits:        16% ( 7380/46121)[GC 276949K->211604K(6209216K), 0.0091803 secs]
[GC 279764K->213453K(6209216K), 0.0108290 secs]
Cleaning commits:        17% ( 7933/46121)[GC 281611K->212910K(6209216K), 0.0071027 secs]
[GC 281070K->214796K(6209216K), 0.0060386 secs]
Cleaning commits:        18% ( 8380/46121)[GC 282956K->217511K(6209216K), 0.0085801 secs]
Cleaning commits:        19% ( 8763/46121)[GC 285671K->219516K(6209216K), 0.0067543 secs]
[GC 287678K->222068K(6209216K), 0.0075882 secs]
[GC 222137K->223144K(6209216K), 0.0054835 secs]
Cleaning commits:        20% ( 9225/46121)[GC 291304K->222667K(6209216K), 0.0050921 secs]
[GC 290827K->223754K(6209216K), 0.0081762 secs]
Cleaning commits:        21% ( 9686/46121)[GC 291914K->225337K(6209216K), 0.0067897 secs]
Cleaning commits:        21% ( 9792/46121)[GC 293497K->227260K(6209216K), 0.0091717 secs]
[GC 295420K->229148K(6209216K), 0.0099031 secs]
Cleaning commits:        21% ( 9968/46121)[GC 297308K->232162K(6209216K), 0.0094873 secs]
Cleaning commits:        21% (10019/46121)[GC 300322K->233982K(6209216K), 0.0105254 secs]
Cleaning commits:        22% (10147/46121)[GC 302142K->236331K(6209216K), 0.0095469 secs]
Cleaning commits:        22% (10223/46121)[GC 304491K->237074K(6209216K), 0.0101256 secs]
[GC 305234K->240100K(6209216K), 0.0105519 secs]
Cleaning commits:        22% (10377/46121)[GC 308260K->242184K(6209216K), 0.0089471 secs]
Cleaning commits:        22% (10518/46121)[GC 310344K->244258K(6209216K), 0.0129478 secs]
Cleaning commits:        23% (10608/46121)[GC 312418K->250378K(6209216K), 0.0149975 secs]
Cleaning commits:        23% (10643/46121)[GC 318538K->249608K(6209216K), 0.0064976 secs]
[GC 317768K->251697K(6209216K), 0.0106580 secs]
Cleaning commits:        23% (10819/46121)[GC 319857K->254422K(6209216K), 0.0104801 secs]
[GC 322582K->256410K(6209216K), 0.0101502 secs]
Cleaning commits:        24% (11070/46121)[GC 324570K->258492K(6209216K), 0.0097239 secs]
Cleaning commits:        24% (11177/46121)[GC 326652K->259942K(6209216K), 0.0127709 secs]
[GC 328102K->262001K(6209216K), 0.0101039 secs]
Cleaning commits:        24% (11307/46121)[GC 330161K->264421K(6209216K), 0.0118688 secs]
Cleaning commits:        24% (11433/46121)[GC 332581K->266010K(6209216K), 0.0081101 secs]
Cleaning commits:        24% (11501/46121)[GC 334170K->267211K(6209216K), 0.0103131 secs]
Cleaning commits:        26% (11992/46121)[GC 335371K->269940K(6209216K), 0.0112067 secs]
Cleaning commits:        26% (12288/46121)[GC 338100K->272652K(6209216K), 0.0126471 secs]
Cleaning commits:        26% (12401/46121)[GC 340812K->275023K(6209216K), 0.0098540 secs]
Cleaning commits:        27% (12473/46121)[GC 343183K->277015K(6209216K), 0.0078316 secs]
Cleaning commits:        27% (12566/46121)[GC 345175K->279245K(6209216K), 0.0118890 secs]
Cleaning commits:        27% (12667/46121)[GC 347405K->280637K(6209216K), 0.0149269 secs]
Cleaning commits:        27% (12744/46121)[GC 348797K->282639K(6209216K), 0.0098945 secs]
Cleaning commits:        27% (12814/46121)[GC 350799K->284370K(6209216K), 0.0116901 secs]
Cleaning commits:        27% (12840/46121)[GC 352530K->286445K(6209216K), 0.0100106 secs]
Cleaning commits:        27% (12877/46121)[GC 354605K->289891K(6209216K), 0.0138893 secs]
Cleaning commits:        28% (12915/46121)[GC 358051K->289834K(6209216K), 0.0077496 secs]
Cleaning commits:        28% (12944/46121)[GC 357994K->291941K(6209216K), 0.0108101 secs]
Cleaning commits:        28% (13004/46121)[GC 360101K->294527K(6209216K), 0.0109895 secs]
Cleaning commits:        28% (13023/46121)[GC 362687K->296599K(6209216K), 0.0100796 secs]
Cleaning commits:        28% (13055/46121)[GC 364759K->298443K(6209216K), 0.0099168 secs]
Cleaning commits:        28% (13088/46121)[GC 366603K->299570K(6209216K), 0.0108803 secs]
Cleaning commits:        28% (13117/46121)[GC 367730K->301613K(6209216K), 0.0112244 secs]
Cleaning commits:        28% (13155/46121)[GC 369773K->302296K(6209216K), 0.0106726 secs]
Cleaning commits:        28% (13189/46121)[GC 370456K->305912K(6209216K), 0.0131232 secs]
Cleaning commits:        28% (13206/46121)[GC 374072K->306111K(6209216K), 0.0127650 secs]
Cleaning commits:        28% (13248/46121)[GC 374271K->307730K(6209216K), 0.0104809 secs]
Cleaning commits:        28% (13264/46121)[GC 375890K->314381K(6209216K), 0.0208222 secs]
Cleaning commits:        28% (13287/46121)[GC 382541K->320775K(6209216K), 0.0209923 secs]

**Comment by donnib** [comment]
also to add, even in the above case, i still have alot of memory left on the machine so it's not a memory bottleneck, it's not disk as that can handle much more read/write and the same goes for the CPU (the CPU is about 50% idle while doing above)

**Comment by rtyley** [comment]
Does the job eventually complete? 20 minutes _is_ abnormally slow - if you look at the timing for cleaning gcc (a repo of similar size), you'll see that the time would be closer to 3 minutes:

https://docs.google.com/spreadsheet/ccc?key=0AsR1d5Zpes8HdER3VGU1a3dOcmVHMmtzT2dsS2xNenc#gid=0

**Comment by donnib** [comment]
Yes it does end in 1.000.0000 ms so roughly 17min.

I tried this on a Windows workstation as well

**Comment by donnib** [comment]
It takes 30seconds to get to 25% percent of the repo so that shows how slow it is after that : 

Idk if this is of any interest (maybe it shows the reason it takes so much time) : 

Commit Tree-Dirt History
------------------------

	Earliest                                                                                                                Latest
	|                                                                                                                                    |
	...............DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDmmm

	D = dirty commits (file tree fixed)
	m = modified commits (commit message or parents changed)
	. = clean commits (no changes to file tree)

Bear in mind this repo has just been cloned from SVN, (idk if that is a deal breaker)

**Comment by donnib** [comment]
@rtyley How can i remove multiple files at the same time with different names ? I read the doc but i can't seem to figure out how to separete between them ?

**Comment by rtyley** [comment]
It's a glob expression, so comma separate inside braces:

```
$ bfg --delete-files '{fileA,fileB,fileC}'  my-repo.git
```

http://stackoverflow.com/a/21636596/438886

**Comment by donnib** [comment]
@rtyley thx, i forgot the single quotes when i tried it, i'll give that a try next time, i did actually a clean and now when i git push i get :  

```
remote: error: refusing to update checked out branch: refs/heads/master
remote: error: By default, updating the current branch in a non-bare repository
remote: error: is denied, because it will make the index and work tree inconsistent
remote: error: with what you pushed, and will require 'git reset --hard' to match
remote: error: the work tree to HEAD.
```

The clone i did was from a local repo (i just migrated from SVN so i have the repo locally) so i did :
` git clone --mirror gitRepo gitrepo1.git`

How do i push the changes safely to the repo ?

**Comment by rtyley** [comment]
At the moment, from your description, I know about 2 copies of your Git repo, both of which exist on your Mac: the original version produced by migrating from SVN, and then the cleaned mirror clone which was made from that. You haven't yet mentioned a remote server, but I presume that's where you're going with this? Let's call that server eddie, with an address of `git@github.com:foo/eddie.git`

Add your remote server eddie as a mirror remote, on your cleaned clone:

```
$ cd /Users/mihai/GitMigration/gitRepo.git
$ git remote add eddie --mirror=push git@github.com:foo/eddie.git
```

Now push to eddie:

```
$ git push eddie
```

You can forget about your SVN version, apart from creating a backup zip of it to ensure you've got a fallback if anything goes wrong.

**Comment by donnib** [comment]
@rtyley thank you, that solved the problem and yes i need to end up with this on a remote server. 

Does BFG support removing things from commit messages ? I need to remove svn infomation from all commits after the migration, like the command `--strip-metadata` does in https://bitbucket.org/atlassian/svn-migration-scripts/overview or see the code here: https://bitbucket.org/atlassian/svn-migration-scripts/src/56bbc7fd48a5b0e32048a62db215055b5da4bd7e/src/main/scala/Clean.scala?at=master

**Comment by rtyley** [comment]
It already does do small rewrites on commit messages (replace old commit-ids with the updated version), but there's no current support to do custom manipulations, other than you writing some Scala code yourself!

Obviously, it's your call, but how many kittens will die if the svn information stays in there?

**Comment by donnib** [comment]
Yeah, i don't think it matters so much, it's just annoying to the eyes to see them when you are reading the commit message of an old commit that was made in SVN.

**Comment by donnib** [comment]
One more question, i did the git push eddie but the branches are not pushed, if i try git push --all i get that --mirror is not compatible, any ideas ?

**Comment by donnib** [comment]
@rtyley after i now have done the clean i am interested in pushing the cleaned copy to our git repo but we are in a middle of a stage now where we have not migrated to Git yet although i would like to keep the Git repo that i have pushed up to date with the SVN changes, because i did the bare clone i lost the svn info, is there a way i can clean the repo, keep the svn info then push it to the remote so i can then run `git svn fetch` and be able to get newer changes to our git repo unti we make the switch ?

So to summarize :

1. I have a cloned Git repo from SVN (not cleaned) which i can do `git svn fetch` on
2. I do a bare clone of above and then use bfg cleaner
3. Add server and push

After point 2 i can't do `git svn fetch` anymore. Is there a way to accomplish this ?

1. Clone svn into a Git repo
2. Clean with BGF 
3. Push to server
4. Fetch new changes from SVN
5. (Optional) Clean again with BFG
6. Push to server new changes

**Comment by donnib** [comment]
@rtyley any ideas on the above ?
